<launch>

  <!-- Filepath for the robot description file; can be overridden -->
  <arg name="model" default="$(find ros_ex)/urdf/imu_filter_compare.urdf.xacro" />

  <!-- Open the URDF, run the xacro preprocessor, and stuff the result into 'robot_description' on the parameter server -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg model)" />

  <!-- Launch some state publishers -->
  <!--<node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />-->
  <!--<node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />-->

  <!-- Run rviz -->
  <node name="rviz" pkg="rviz" type="rviz" required="true" />

  <!-- My TF publisher -->
  <node name="tf_publish" pkg="ros_ex" type="tf_publish" required="true" />

  #### IMU Driver ###########################################################

  <node pkg="phidgets_imu" type="phidgets_imu_node" name="phidgets">

    <param name="frame_id" value="imu_link" /> <!-- doesn't matter, we'll provide our own tf -->

    # supported data rates: 4 8 16 24 32 40 ... 1000 (in ms)
    <param name="period" value="8" />

    # optional param serial_number, default is -1
    <!-- <param name="serial_number" value="123456" /> -->

    # compass correction params (see http://www.phidgets.com/docs/1044_User_Guide)
    <param name="cc_mag_field" value="0.618421" />
    <param name="cc_offset0" value="0.066053" />
    <param name="cc_offset1" value="0.207136" />
    <param name="cc_offset2" value="-0.112013" />
    <param name="cc_gain0" value="1.526209" />
    <param name="cc_gain1" value="1.515731" />
    <param name="cc_gain2" value="1.809127" />
    <param name="cc_t0" value="-0.066839" />
    <param name="cc_t1" value="-0.014559" />
    <param name="cc_t2" value="-0.061476" />
    <param name="cc_t3" value="0.012267" />
    <param name="cc_t4" value="-0.015864" />
    <param name="cc_t5" value="0.014241" />

  </node>

  #### madgwick, stateless ###############################################

  <node pkg="imu_filter_madgwick" type="imu_filter_node" name="stateless">

    <param name="use_mag" value="true" /> <!-- Use magnetic field data -->
    <param name="use_magnetic_field_msg" value="true" /> <!-- Use the new msg type -->
    <param name="publish_tf" value="false" /> <!-- Do not publish a transform, orca_base will do that -->
    <param name="world_frame" value="nwu" /> <!-- to match complementary -->

    <param name="stateless" value="true" />

    <remap from="imu/data" to="imu/data_stateless"/>

  </node>

  #### madgwick ###############################################

  <node pkg="imu_filter_madgwick" type="imu_filter_node" name="madgwick">

    <param name="use_mag" value="true" /> <!-- Use magnetic field data -->
    <param name="use_magnetic_field_msg" value="true" /> <!-- Use the new msg type -->
    <param name="publish_tf" value="false" /> <!-- Do not publish a transform, orca_base will do that -->
    <param name="world_frame" value="nwu" /> <!-- to match complementary -->

    <remap from="imu/data" to="imu/data_madgwick"/>

  </node>

  #### complementary ###############################################

  <node pkg="imu_complementary_filter" type="complementary_filter_node" name="complementary">

    <param name="use_mag" value="true" /> <!-- Use magnetic field data -->
    <param name="publish_tf" value="false" /> <!-- Do not publish a transform, orca_base will do that -->

    <remap from="imu/data" to="imu/data_complementary"/>

  </node>

</launch>
